### ISSUE TITLE
Improvements to autograd profiler for broad usefulness as a primary profiler

### ISSUE THREAD

**Issue by dzhulgakov** [pytorch/pytorch#23980]
## 🚀 Feature
Listing small improvements that might make autograd profiler more broadly useful.

## Motivation

Profiling is generally useful in many cases both in training and production inference deployment. Autograd profiler is great but it's a bit non-obvious how to use it outside of python environment

## Suggested features

Here's my list of suggestions, please provide more ideas

### API and usability

- [ ] Move namespace from `torch.autograd.profiler` to `torch.profiler` to better reflect purpose (as it's not autograd specific and causes confusion - e.g. #15857)
- [ ] Expose and document `RecordFunction` call in both Python and C++ and encourage users to instrument their own code with additional ranges (for example, model runs in inference). #19422 adds it to Python but it's still not obvious in C++
- [ ] Move all post-processing of events, including chrome tracing export, to C++ in order to enable inference-only use cases (#16580 is a start, but there's more to do, #18630 for documentation)
- [ ] Write a comprehensive tutorial covering all of the features

### Features
- [ ] Instrument TorchScript Executor with ranges for each Function and potentially Instruction (with pointers back to the source code line and number). This would be useful for Icicle-style charts
- [ ] Preserve scope nesting when crossing thread boundaries with `_fork()` or `backward()` execution. (approach similar to #22365 can be used)
- [ ] Consider adding structured metadata for each scope/range in RecordFunction in addition to simple string name (might be useful to track things like model id)
- [ ] Make profiler enabling thread-safe such that it can be attached to the entire process with other threads running PyTorch code without explicit synchronization. (might be hard to do - see #19420 for discussion)

### Metrics
- [ ] Capture more hardware counters for per-operator execution: FLOPs executed, floating point exceptions, memory bandwidth utilized (might be hard to do generally, but available on some platforms)
- [ ] Attach analytical FLOPs estimates to individual operators in ATen to cross-compare achieved utilizes with the roofline model
- [ ] Capture and track intra-op parallel op execution on CPU (so that it's easy to spot whether parallelization happened or not). Ideally track individual subtasks execution on the chrome trace profile too to visualize overloading of a intra-op threadpool.

**Comment by salexspb** [comment]
Regarding python vs c++ analysis of traces - it is not really blocking inference. Inference benchmarks runs are scheduled from python while the executor is in c++. I agree that there could be a scenario though when c++ only runtime is needed. For example, many models running in parallel at inference time and someone willing to do an analysis of what is going on there. Though, in my cases, this can be done using scuba logging as well.

**Comment by jamesr66a** [comment]
> Move namespace from torch.autograd.profiler to torch.profiler to better reflect purpose (as it's not autograd specific and causes confusion - e.g. #15857)

It actually is autograd-specific in that the profiling hooks are in the VariableType wrappers. Other types of operators, such as `prim::` operators, are not recorded because they do not participate in autograd. I think it's more accurate to rephrase this as actually making the profiler more general beyond autograd ops

**Comment by ngimel** [comment]
Closing, mostly done.
